File: 51-n-queens/n-queens.py
# Brute Force Code & Optimal Code
class Solution:
    def __init__ (self):
        # list to store the all valid board configuration
        self.result = []
    # backtracking function to place queens row by rows
    def solve(self, board: list[list[str]], row: int, columns: set[int], diagonals: set[int], anti_diagonals: set[int]):
        # if queens have been placed in all rows store the current board configurations
        if row == len(board):
            self.result.append(["".join(r) for r in board])
            return 
        # try placing a queen in every column of the current rows
        for col in range(len(board)):
            # compute the current diagonal identifiers
            diagonal_id = row - col
            antidiagonal_id = row + col
            # skip this position if it is already under attack
            if (col in columns or diagonal_id in diagonals or antidiagonal_id in anti_diagonals):
                continue
            # mark the current column and diagonal as occupieds
            columns.add(col)
            diagonals.add(diagonal_id)
            anti_diagonals.add(antidiagonal_id)
            # place the queens
            board[row][col] = "Q"
            # recursively for the next rows
            self.solve(board, row + 1, columns, diagonals, anti_diagonals)
            # remove the queens and free the columns and diagonals
            columns.remove(col)
            diagonals.remove(diagonal_id)
            anti_diagonals.remove(antidiagonal_id)
            board[row][col] = "."
    # return all valid N-Queens configurations 
    def solveNQueens(self, n: int) -> List[List[str]]:
        # if the boards size is zeros, return any empty list
        if n == 0:
            return []
        # clear any previous stored solutions
        self.result = []
        # create an empty n * n solutions
        board = [["." for _ in range(n)] for _ in range(n)]
        # sets to keep track of occupied columns main diagonals and anti-diagonals
        columns = set()
        diagonals = set()
        anti_diagonals = set()
        # start placing queens from first rows
        self.solve(board, 0, columns, diagonals, anti_diagonals)
        # return all valid board configurations
        return self.result

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 51-n-queens/n-queens.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    // vector to store all valid board configurations
    vector<vector<string>> result;
    // backtracking function to place queens row by row
    void solve(vector<string>& board, int row, unordered_set<int>& columns,unordered_set<int>& diagonals, unordered_set<int>& anti_diagonals) {
        // if queens have been placed in all rows store the current board configuration
        if (row == board.size()) {
            // store the current board
            result.push_back(board);
            return;
        }
        // try placing a queen in every column of the current row
        for (int col = 0; col < board.size(); col++) {
            // compute the current diagonal identifiers
            int diagonal_id = row - col;
            int antidiagonal_id = row + col;
            // skip this position if it is already under attack
            if (columns.count(col) || diagonals.count(diagonal_id) || anti_diagonals.count(antidiagonal_id)) {
                continue;
            }
            // mark the current column and diagonals as occupied
            columns.insert(col);
            diagonals.insert(diagonal_id);
            anti_diagonals.insert(antidiagonal_id);
            // place the queen
            board[row][col] = 'Q';
            // recursively place queens in the next row
            solve(board, row + 1, columns, diagonals, anti_diagonals);
            // remove the queen and free the column and diagonals
            columns.erase(col);
            diagonals.erase(diagonal_id);
            anti_diagonals.erase(antidiagonal_id);
            board[row][col] = '.';
        }
    }
    // return all valid N-Queens configurations
    vector<vector<string>> solveNQueens(int n) {
        // if board size is zero return an empty list
        if (n == 0) {
            return {};
        }
        // clear any previously stored solutions
        result.clear();
        // create an empty n x n board
        vector<string> board(n, string(n, '.'));
        // sets to keep track of occupied columns main diagonals, and anti-diagonals
        unordered_set<int> columns;
        unordered_set<int> diagonals;
        unordered_set<int> anti_diagonals;
        // start placing queens from the first row
        solve(board, 0, columns, diagonals, anti_diagonals);
        // return all valid board configurations
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)